### PR TITLE
Add adjust window scaling function for display preferences

### PR DESCRIPTION
**Pull request by zhuyaliang**
This PR adds a new feature to the display preferences interface,This new feature adjusts the scaling of the window.
But please note that the adjustment of the window scale is real-time. When you select the scale, the window will start to change immediately

### test
```
mate-display-properties
adjust window scaling
```
2021-03-04 10-49-05屏幕截图

### CODE REVIEW

File: capplets/display/xrandr-capplet.c [hunk 1]
```diff
@@ -452,6 +457,52 @@ mirror_screens_is_supported (App *app)
...
+static void
+on_scale_btn_active_changed_cb (GtkWidget *widget,
+                                App       *app)
+{
+    gint scale;
+
+    if (!gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (widget)))
+        return;
+
+    scale = g_object_get_data (G_OBJECT (widget), "scale");
```
Comment (rbuj) [inline capplets/display/xrandr-capplet.c:469]:
```
xrandr-capplet.c:469:11: warning: assignment to 'gint' {aka 'int'} from 'gpointer' {aka 'void *'} makes integer from pointer without a cast [-Wint-conversion]
```

File: capplets/display/xrandr-capplet.c [hunk 2]
```diff
@@ -452,6 +457,52 @@ mirror_screens_is_supported (App *app)
...
+    {
+        scale_btn = gtk_radio_button_new_with_label_from_widget (group, button_label[i]);
+        if (!group)
+            group = GTK_RADIO_BUTTON (scale_btn);
+        if (i == scale)
+            gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (scale_btn), TRUE);
+
+        g_object_set_data (G_OBJECT (scale_btn),
+                          "scale",
+                           i);
```
Comment (rbuj) [inline capplets/display/xrandr-capplet.c:495]:
```
xrandr-capplet.c:495:28: warning: passing argument 3 of 'g_object_set_data' makes pointer from integer without a cast [-Wint-conversion]
```

**Review by rbuj — COMMENTED** [review]
There are two int-conversion warnings:

File: capplets/display/xrandr-capplet.c [hunk 3]
```diff
@@ -452,6 +457,52 @@ mirror_screens_is_supported (App *app)
...
+static void
+rebuild_scale_window (App *app)
+{
+    GtkRadioButton *group = NULL;
+    GtkWidget      *scale_btn;
+    gint32          scale;
+    int             i;
+    const char     *button_label[] = {_("auto detect"), "100%", "200%", NULL};
+
+    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), (GtkCallback) gtk_widget_destroy, NULL);
```
Comment (rbuj) [inline capplets/display/xrandr-capplet.c:482]: New build warning
```
xrandr-capplet.c:482:61: warning: cast between incompatible function types from 'void (*)(GtkWidget *)' {aka 'void (*)(struct _GtkWidget *)'} to 'void (*)(GtkWidget *, void *)' {aka 'void (*)(struct _GtkWidget *, void *)'} [-Wcast-function-type]
```

Comment (zhuyaliang) [reply capplets/display/xrandr-capplet.c]: The warning has been fixed

File: capplets/display/xrandr-capplet.c (hunk 2 above)
Comment (zhuyaliang) [reply capplets/display/xrandr-capplet.c]: The warning has been fixed

Comment (zhuyaliang) [reply capplets/display/xrandr-capplet.c]: The warning has been fixed

**Review by lukefromdc — APPROVED** [review]
I only saw deprecation warnings building this after the earlier fixes. Works quite nicely, and as expected can even be invoked from the xrandr panel applet by selecting "configure display settings" which calls the capplet directly. Autodetect has never worked with any of my monitor/hookup/video card settings at any time (no change here) but manual selection works perfectly.

This is a real usability improvement for end users with hidpi screens and no experience using dconf-editor.

**Comment by lukefromdc** [comment]
I also looked over memory management and didn't see any obvious leaks, though as always I could be wrong there

**Comment by raveit65** [comment]
@lukefromdc 
Maybe you like to test this nice work from too?
https://github.com/zhuyaliang/mate-recorder/issues/1

**Review by rbuj — COMMENTED** [review]
In addition to the changes I suggest below, I think it's better to use a combobox to save space as this dialog has a compact design. It's much simpler than using an array of radio buttons.
```diff
diff --git a/capplets/display/xrandr-capplet.c b/capplets/display/xrandr-capplet.c
index 5836281..67c83c9 100644
--- a/capplets/display/xrandr-capplet.c
+++ b/capplets/display/xrandr-capplet.c
@@ -470,20 +470,17 @@ on_scale_btn_active_changed_cb (GtkWidget *widget,
     g_settings_set_int (app->scale_settings, WINDOW_SCALE_KEY, scale);
 }
 
-static void foreach_destroy_cb (GtkWidget *child, gpointer data)
-{
-    gtk_widget_destroy (child);
-}
-
 static void
 rebuild_scale_window (App *app)
 {
     GtkRadioButton *group = NULL;
     gint32          scale;
     int             i;
-    const char     *button_label[] = {_("auto detect"), "100%", "200%", NULL};
+    const char     *button_label[] = {_("Auto detect"), _("100%"), _("200%"), NULL};
+
+    /* destroy any previous scale_btn */
+    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), G_CALLBACK (gtk_widget_destroy), NULL);
 
-    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), (GtkCallback) foreach_destroy_cb, NULL);
     scale = g_settings_get_int (app->scale_settings, WINDOW_SCALE_KEY);
 
     for (i = 0; button_label[i] != NULL; i++)
```
### Before
Screenshot at 2021-03-05 10-19-38
### After
Screenshot at 2021-03-05 10-17-54

Edit: The box order should also be rearranged as follows:
Screenshot at 2021-03-05 10-35-54

**Comment by zhuyaliang** [comment]
@rbuj ```gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), G_CALLBACK (gtk_widget_destroy), NULL)``` 
This operation will generate compilation warnings, which was the case in my previous version
>  I think it's better to use a combobox to save space as this dialog has a compact design. It's much simpler than using an array of radio buttons.

I thought so at the beginning, but there will be problems in operation。When you change the window scale, the drop-down window is not fully displayed
Like this：

You can do that again
In the case of 100% scaling, first click the combobox component, such as resolution
Then adjust the scale to 200%, click the combobox component again, and you will find the problem

File: capplets/display/xrandr-capplet.c (hunk 3 above)
Comment (zhuyaliang) [reply capplets/display/xrandr-capplet.c]: Please confirm whether to change this problem

**Comment by rbuj** [comment]
> @rbuj `gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), G_CALLBACK (gtk_widget_destroy), NULL)`
> This operation will generate compilation warnings, which was the case in my previous version

It removes the warning below:
```
xrandr-capplet.c:473:60: warning: unused parameter 'data' [-Wunused-parameter]
```

**Comment by rbuj** [comment]
> > I think it's better to use a combobox to save space as this dialog has a compact design. It's much simpler than using an array of radio buttons.
> 
> I thought so at the beginning, but there will be problems in operation。When you change the window scale, the drop-down window is not fully displayed
> Like this：
> 
> You can do that again
> In the case of 100% scaling, first click the combobox component, such as resolution
> Then adjust the scale to 200%, click the combobox component again, and you will find the problem

I couldn't reproduce you issue, perhaps it's due to another external factor.

File: capplets/display/xrandr-capplet.c [hunk 4]
```diff
@@ -452,6 +457,54 @@ mirror_screens_is_supported (App *app)
...
+rebuild_scale_window (App *app)
+{
+    GtkRadioButton *group = NULL;
+    gint32          scale;
+    int             i;
+    const char     *button_label[] = {_("auto detect"), _("100%"), _("200%"), NULL};
+
+    /* destroy any previous scale_btn */
+    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), G_CALLBACK (gtk_widget_destroy), NULL);
+    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), (GtkCallback) foreach_destroy_cb, NULL);
```
Comment (rbuj) [inline capplets/display/xrandr-capplet.c:483]: foreach_destroy_cb doen't exist now.

**Comment by zhuyaliang** [comment]
@rbuj I changed the code as you suggested，But I still decided to use ```radio buttons``` because for whatever reason, ```combo boxes``` have problems with me，Hope to be understood

**Comment by rbuj** [comment]
> @rbuj I changed the code as you suggested，But I still decided to use `radio buttons` because for whatever reason, `combo boxes` have problems with me，Hope to be understood

Ok, Please edit the display-capplet.ui file using glade or use the attachment instead, it will also update the property names:

[display-capplet.ui.gz](https://github.com/mate-desktop/mate-control-center/files/6089860/display-capplet.ui.gz)

File: capplets/display/xrandr-capplet.c (hunk 4 above)
Comment (rbuj) [reply capplets/display/xrandr-capplet.c]: You deleted the wrong line.

Comment (zhuyaliang) [reply capplets/display/xrandr-capplet.c]: It's my fault. I'm in a hurry to go home for dinner

**Comment by rbuj** [comment]
The scaling factor is not a setting that can be set individually for a monitor and it doesn't change if one of the monitors is turned off. You can find other ways to implement it.
```diff
diff --git a/capplets/display/display-capplet.ui b/capplets/display/display-capplet.ui
index aff649d..be905a8 100644
--- a/capplets/display/display-capplet.ui
+++ b/capplets/display/display-capplet.ui
@@ -464,7 +464,7 @@
               </packing>
             </child>
             <child>
-              <object class="GtkBox" id="vbox5">
+              <object class="GtkBox" id="scale_vbox">
                 <property name="visible">True</property>
                 <property name="can-focus">False</property>
                 <property name="orientation">vertical</property>
@@ -486,17 +486,7 @@
                   </packing>
                 </child>
                 <child>
-                  <object class="GtkButtonBox" id="scale_bbox">
-                    <property name="visible">True</property>
-                    <property name="can-focus">False</property>
-                    <property name="valign">center</property>
-                    <property name="layout-style">expand</property>
-                  </object>
-                  <packing>
-                    <property name="expand">False</property>
-                    <property name="fill">False</property>
-                    <property name="position">1</property>
-                  </packing>
+                  <placeholder/>
                 </child>
               </object>
               <packing>
diff --git a/capplets/display/xrandr-capplet.c b/capplets/display/xrandr-capplet.c
index fadc78a..1e859ec 100644
--- a/capplets/display/xrandr-capplet.c
+++ b/capplets/display/xrandr-capplet.c
@@ -62,6 +62,7 @@ struct App
     GtkWidget     *rotation_combo;
     GtkWidget     *panel_checkbox;
     GtkWidget     *scale_bbox;
+    GtkWidget     *scale_vbox;
     GtkWidget     *clone_checkbox;
     GtkWidget     *show_icon_checkbox;
     GtkWidget      *primary_button;
@@ -479,7 +480,13 @@ rebuild_scale_window (App *app)
     const char     *button_label[] = {_("auto detect"), _("100%"), _("200%"), NULL};
 
     /* destroy any previous scale_btn */
-    gtk_container_foreach (GTK_CONTAINER (app->scale_bbox), G_CALLBACK (gtk_widget_destroy), NULL);
+    if (app->scale_bbox != NULL)
+        return;
+
+    app->scale_bbox = gtk_button_box_new (GTK_ORIENTATION_HORIZONTAL);
+    gtk_container_set_border_width (GTK_CONTAINER (app->scale_bbox), 6);
+    gtk_container_add (GTK_CONTAINER (app->scale_vbox), app->scale_bbox);
+
     scale = g_settings_get_int (app->scale_settings, WINDOW_SCALE_KEY);
 
     for (i = 0; button_label[i] != NULL; i++)
@@ -500,8 +507,8 @@ rebuild_scale_window (App *app)
                           G_CALLBACK (on_scale_btn_active_changed_cb),
                           app);
         gtk_container_add (GTK_CONTAINER (app->scale_bbox), scale_btn);
-        gtk_widget_show (scale_btn);
     }
+    gtk_widget_show_all (app->scale_bbox);
 }
 
 static void
@@ -2487,7 +2494,7 @@ run_application (App *app)
     g_signal_connect (app->rotation_combo, "changed",
                      G_CALLBACK (on_rotation_changed), app);
 
-    app->scale_bbox = _gtk_builder_get_widget (builder, "scale_bbox");
+    app->scale_vbox = _gtk_builder_get_widget (builder, "scale_vbox");
 
     app->clone_checkbox = _gtk_builder_get_widget (builder, "clone_checkbox");
     g_signal_connect (app->clone_checkbox, "toggled",
```

**Comment by zhuyaliang** [comment]
@rbuj Why not implement it in UI file？

**Comment by zhuyaliang** [comment]
@rbuj Why not implement it in UI file？

**Comment by rbuj** [comment]
What do you want to implement in the ui file? Are you talking about scale_bbox or radio buttons? With the changes I propose, the scale button box and the radio buttons are only initialized once, there is no need to destroy the children of the widget (GtkButtonBox) or the widget itself (GtkRadioButtons).

**Comment by zhuyaliang** [comment]
@rbuj Code has been updated, please check

**Comment by rbuj** [comment]
@zhuyaliang Thank you so much for adding this improvement.

**Pull request merged**